### ISSUE TITLE
[Feature] Allow calling in cluster services

### ISSUE THREAD

**Issue by samj1912** [kyverno/KDP#14]
### Problem Statement

There is currently no easy way to extend kyverno capabilities. Allowing kyverno to make calls to in-cluster services could unlock such capabilities.

### Solution Description

Allow a new `serviceCall` context variable type.

```yaml
rules:
- name: example-service-call
  context:
  # name of the variable to capture the service output from
  - name: output
    serviceCall:
      service:
        name: service-name
        namespace: some-namespace
        # optional
        port: 80
        # optional
        path: validate-namespace
        # optional url GET params
        params:
          namespace: '{{ request.object.metadata.namespace }}'
        # optional JSON post data. If defined.
        data:
          method: '{{ request.method }}'
        # http verb for the call. One of the http verb names. Defaults to GET if data is not defined otherwise POST
        verb: POST
      # optional timeout in seconds per retry, default is 2s
      timeout: 10
      # optional retries. Defaults to 1
      retries: 2
      # optional jmespath expression evaluated on the response before the variable is assigned 
      jmesPath: "items | length(@)"
```

The above service calls would limit the return accept type to json only. The encoding would also be json for the post data. 

We would require an in cluster `Service` object to be defined to reach out to.

### Research

- [X] I have read and followed the documentation AND the [troubleshooting guide](https://kyverno.io/docs/troubleshooting/).
- [X] I have searched other issues in this repository and mine is not recorded.

**Comment by chipzoller** [comment]
Do you want to covert this to a KDP? Would you plan on developing this for 1.8.0 potentially?

**Comment by Dentrax** [comment]
Dropping some use case about the proposal as answering of @JimBugwadia's comment: https://github.com/kyverno/kyverno/pull/3820#issuecomment-1191710517

> @Dentrax - are you looking for this to address a use case? If so, I would like to learn more about that.

Image you are running hundreds of clusters and want to create some _excluding_ policies using _ConfigMap_. For example, we want to exclude some critical images such as _istio_, _node-local-dns_ etc. But this task requires some stuff such creating ConfigMaps in every cluster, and we have to keep maintaining those.

I'd like to store my _context_ in a centralized config solution such as _consul_. By doing so, we are able to store all excluding rules in a single place for better maintainability and real-time operation for _upserting_ the new configs.

But I just noticed this issue is for _in-cluster services_ only. What I was looking for is that _external data_. (I actually mean _out-cluster_ arbitrary apiCall) Should we create another KDP for this?

```
rules:
- name: get-excluded-images
  context:
  - name: excludedImages
    apiCall:
      urlPath: "https://consul.test.com:8500/v1/kv/exluded-images"
```

cc @developer-guy

**Comment by JimBugwadia** [comment]
@Dentrax - thanks for the example use case! 

In-cluster, or an external API, should work the same. Both will need to be secured. 

So, going along with your Consul example, how would you specify the encryption and authentication / authorization for the Consul API?
- Encryption: provide a certificate / root CA?
- Authentication: username and password? token? cert? something else?

In either case, the network policy would also need to be configured to allow this call.

**Comment by zswanson** [comment]
I'd be interested in leveraging this with an internal 'metadata' service we have that our application devs populate about each app during application onboarding. This contains a lot traceability data such as team ownership, cost allocation data, etc. We have a desire to allow/deny certain kubernetes attributes based on that metadata, such as allowing an Ingress to declare a 'high priority' ingressClassName only if its owned by certain groups (and actually validating that ownership through the service). The metadata service is hosted within a central kubernetes cluster so we'd need both in-cluster and external-cluster calls to it.

**Comment by chipzoller** [comment]
@zswanson how would you expect to handle authN/authZ to these external services? Do you have requirements there?

**Comment by zswanson** [comment]
@chipzoller Looks like this particular internal service doesn't require any auth for GET right now. The rest of the service (which we wouldn't need to use) are Bearer token JWT. I think generally our in-house stuff uses username/password or JWT tokens.

**Comment by eddycharly** [comment]
Isn't it going to be tricky to implement in CLI when trying to validate manifests against rules without deploying ?

**Comment by aslafy-z** [comment]
@eddycharly afaik, mocking of the resulting context values exists in the test framework, there's no mock of the http request in itself.

**Comment by chipzoller** [comment]
Why don't we start simple with this feature (and put it behind an alpha feature flag) by getting the basics down such as:

* HTTP only
* No custom headers
* GET method
* JSON response only

**Comment by chipzoller** [comment]
Closing this issue since the KDP was approved and a [PR](https://github.com/kyverno/kyverno/pull/5755) has already been merged, tracking for release in Kyverno 1.10.
